### ISSUE TITLE
Fix intermittent failure of TestLeaseManagement.

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#184]
There is some race condition, which needs investigation. I have seen it failing locally and on Jenkins.

**Comment by shvachko** [comment]
Stack trace from a local failure:

```
Running org.apache.giraffa.TestLeaseManagement
Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 42.86 sec <<< FAILURE! - in org.apache.giraffa.TestLeaseManagement
testLeaseMigration(org.apache.giraffa.TestLeaseManagement)  Time elapsed: 19.42 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <FileLease[holder=DFSClient_NONMAPREDUCE_1048738879_1, path=/testLeaseFailure, lastUpdate=1436309675712]>
     got: <FileLease[holder=DFSClient_NONMAPREDUCE_1048738879_1, path=/testLeaseFailure, lastUpdate=1436309669176]>

    at org.junit.Assert.assertThat(Assert.java:778)
    at org.junit.Assert.assertThat(Assert.java:736)
    at org.apache.giraffa.TestLeaseManagement.testLeaseMigration(TestLeaseManagement.java:270)
```

Another one from Jenkins

```
Running org.apache.giraffa.TestLeaseManagement
Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 31.169 sec <<< FAILURE! - in org.apache.giraffa.TestLeaseManagement
testLeaseRecovery(org.apache.giraffa.TestLeaseManagement)  Time elapsed: 5.042 sec  <<< FAILURE!
java.lang.AssertionError:
Expected: is null
     got: <FileLease[holder=DFSClient_NONMAPREDUCE_369262095_1, path=/testLeaseRecovery, lastUpdate=1435994039179]>

        at org.junit.Assert.assertThat(Assert.java:778)
        at org.junit.Assert.assertThat(Assert.java:736)
        at org.apache.giraffa.TestLeaseManagement.testLeaseRecovery(TestLeaseManagement.java:180)
```

**Comment by weilintsaiWand** [comment]
I spent some time on it, trying to figure it by tracing code but still have no clues. 
My first idea was to hack the code and let the assertion always trigger but in vain.
Then I decided to use brute-force (stupid) approach to grasp more information.
Hence, I use the same [approach to run tests multiple times](http://stackoverflow.com/questions/1492856/easy-way-of-running-the-same-junit-test-over-and-over)<br>
Here are some experimental results:
- Commented out all other tests to run `testLeaseRecovery` only for 5000 times.
  The failure rate is 3/5000.
- Commented out all other tests to run `testLeaseRecovery` and `testLeaseMigration` for 500 times. <br>
  The failure rate for `testLeaseRecovery` is 16/500<br>
  The failure rate for `testLeaseMigration` is 1/500<br>
  Besides, I found the third kind of syndrome.

```
    java.lang.AssertionError:
    Expected: is <CLOSED>
    got: <UNDER_CONSTRUCTION>

    at org.junit.Assert.assertThat(Assert.java:778)
    at org.junit.Assert.assertThat(Assert.java:736)
at org.apache.giraffa.TestLeaseManagement.testLeaseRecovery(TestLeaseManagement.java:178)
```

There's tenfold failure rate difference. Hence, I was wondering if there's a thread created by `testLeaseMigration` which affect `testLeaseRecovery`. However, we still need more investigation.

By the way, if we comment out `getNSTable().put(put)` in `INodeManager.updateINodeLease()` and run  `testLeaseRecovery` only. The failure rate is 0/12500.

**Comment by weilintsaiWand** [comment]
More investigation result, seems like the failures come from different reason:

###### Assert Null failure

```
java.lang.AssertionError:
Expected: is null
     got: <FileLease[holder=DFSClient_NONMAPREDUCE_369262095_1, path=/testLeaseRecovery, lastUpdate=1435994039179]>

        at org.junit.Assert.assertThat(Assert.java:778)
        at org.junit.Assert.assertThat(Assert.java:736)
        at org.apache.giraffa.TestLeaseManagement.testLeaseRecovery(TestLeaseManagement.java:180)
```

It's triggered under a race condition between `renewLease()` and `internalReleaseLease()`. After `internalReleaseLease()` automatically recover the file lease, a background running `renewLease()` updates the lease to Non-NULL value by calling `updateINodeLease()`. The solution is to check if to-be-update-lease node is closed in prePut. The patch pass 400 simulations.

## 

###### Assert CLOSED failure

```
 java.lang.AssertionError:
    Expected: is <CLOSED>
    got: <UNDER_CONSTRUCTION>

    at org.junit.Assert.assertThat(Assert.java:778)
    at org.junit.Assert.assertThat(Assert.java:736)
at org.apache.giraffa.TestLeaseManagement.testLeaseRecovery(TestLeaseManagement.java:178)
```

Looks like a true issue. The assert get UNDER_CONSTRUCTION because `internalReleaseLease()` never take place. Need further investigation.

## 

###### Lease migration failure

Haven't investigated yet.

**Comment by weilintsaiWand** [comment]
Also found a potential risky code in `NamespaceProcessor.java`

``` java
public boolean internalReleaseLease(FileLease lease, String src)
      throws IOException {
    LOG.info("Recovering " + lease + ", src=" + src);

    try {
      INode iNode = nodeManager.getINode(src);
      if(iNode.getBlocks().size() == 0) {
        LOG.info("Zero blocks detected.");
        return false;
      }
      if(iNode.getFileState().equals(FileState.RECOVERING)) {
        LOG.info("File under recovery already.");
        return false;
      }
      iNode.setState(FileState.RECOVERING);
      nodeManager.updateINode(iNode, BlockAction.RECOVER);
    } catch (IOException e) {
      LOG.error(e);
      return false;
    }
    return true;
}
```

The second `if` statement is never true since every time we set `iNode.setState(FileState.RECOVERING);` the file state will be changed to `CLOSED` in `BlockManagementAgent` like 

``` java
updateField(kvs, FileField.FILE_STATE,
        Bytes.toBytes(FileState.CLOSED.toString()));
```

**Comment by shvachko** [comment]
Some interesting investigation.
Just keep in mind that race conditions could be tricky, so if you feel stuck ask questions.

**Comment by weilintsaiWand** [comment]
Yes the condition is tricky and I do get stuck now. Just summarize what I have as below:
First, there's a cc726c0a50f87f8944f8511ead6d6f53b23829f7 commit in my own repository .
There are three types of failures:
1. `Assert Null failure` from `testLeaseRecovery`
2. `Assert CLOSED failure` from `testLeaseRecovery`
3. `Lease migration failure`

This patch can 
1. Solve case 1 completely (Pass more than 20000 times simulation)
2. Solve case 2 when we run `testLeaseRecovery` only (Pass more than 20000 times simulation)

But can not 
1. Solve case 3 
2. Solve case 2 if we runs `testLeaseRecovery` and `testLeaseMigration` together 

## 

### How it works

###### To solve case 1

Case one comes from race condition between `renewLease()` and `internalReleaseLease()`. As a result, I added 

``` java
if (checkIfRacing(kvs, e, put)) { return; }
```

in `prePut()`

###### To solve case 2

Case two might due to some uncaught exception in recovery. Hence, I added `recoverLastBlockWrapper(kvs, e, put)` in `prePut()` which looks likes

``` java
  private void recoverLastBlockWrapper(List<Cell> kvs,
                                ObserverContext<RegionCoprocessorEnvironment> e,
                                Put put) {
    putFileState(put, e, FileState.RECOVERING);
    try {
      recoverLastBlock(kvs);
    } catch (Exception exception) {
        putFileState(put, e, FileState.UNDER_CONSTRUCTION);
    }
  }
```

and refine code in `recoverLastBlock()` to throw exception whenever abnormal case happens.

## 

### The problem now

If we runs `testLeaseRecovery` only, it would pass 20000 runs without any error. However, if we
runs  `testLeaseRecovery` and `testLeaseMigration` together, both case 2 and case 3 will be triggered 

##### Possible cause

Basically I have no idea now. The only possibility I can image is the we MIGHT not clear/reset the environment completely in @after method for `testLeaseMigration`. As a result, if we `testLeaseRecovery` and `testLeaseMigration` together, something bad in testLeaseMigration will impact the operation in `testLeaseRecovery`

## 

### Update 1

I add sleep in @After as

``` java
  @After
  public void after() throws IOException {
    IOUtils.cleanup(LOG, grfs, nodeManager, connection);
    try {
      Thread.sleep(10000L);
    } catch (Exception ignored) {}
  }
```

And there's no failure for 250 runs. Though that's a small sample ...

**Comment by weilintsaiWand** [comment]
As Konstantin suggested, create #196 to deal with first type of failure

**Comment by weilintsaiWand** [comment]
Quick update for some experimental results:

##### Code version:

Latest patch with 1. `checkIfRacing()` and 2. `recoverLastBlockWrapper()` in `prePut`(). 

##### Test to run:

All tests in `TestLeaseManagement`except `testLeaseMigration`. In other words, we ran `testLeaseCreation()`, `testLeaseFailure()`, `testLeaseRecovery()`, `testClientLeaseRecovery()` simultaneously.  

##### How many runs:

3500 times for each tests. 14,000 times in total.

##### Result

No error.
